pkg/controller/messaginginfra/messaginginfra_controller.go

```diff
@@ -0,0 +1,216 @@
+	certController := cert.NewCertController(mgr.GetClient(), mgr.GetScheme(), 24*30*time.Hour, 24*time.Hour)
+	brokerController := broker.NewBrokerController(mgr.GetClient(), mgr.GetScheme(), certController)
+	routerController := router.NewRouterController(mgr.GetClient(), mgr.GetScheme(), certController)
+	return &ReconcileMessagingInfra{
+		client:           mgr.GetClient(),
+		certController:   certController,
+		routerController: routerController,
+		brokerController: brokerController,
+		reader:           mgr.GetAPIReader(),
+		scheme:           mgr.GetScheme(),
+		namespace:        namespace,
+	}
+}
+
+func add(mgr manager.Manager, r *ReconcileMessagingInfra) error {
+
+	c, err := controller.New("messaginginfra-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	err = c.Watch(&source.Kind{Type: &v1beta2.MessagingInfra{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	err = c.Watch(&source.Kind{Type: &appsv1.StatefulSet{}}, &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &v1beta2.MessagingInfra{},
+	})
+	return err
+}
+
+func (r *ReconcileMessagingInfra) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling MessagingInfra")
+
+	ctx := context.TODO()
+
+	found := &v1beta2.MessagingInfra{}
+	err := r.client.Get(ctx, request.NamespacedName, found)
+	if err != nil {
+		if k8errors.IsNotFound(err) {
+			reqLogger.Info("MessagingInfra resource not found. Ignoring since object must be deleted")
+			return reconcile.Result{}, nil
+		}
+		reqLogger.Error(err, "Failed to get MessagingInfra")
+		return reconcile.Result{}, err
+	}
+
+	ic := &infraContext{
+		ctx:    ctx,
+		client: r.client,
+		infra:  found,
+		status: found.Status.DeepCopy(),
+	}
+
+	// Initialize status
+	err = ic.ProcessSimple(func(infra *v1beta2.MessagingInfra) error {
+		if infra.Status.Phase == "" || infra.Status.Phase == v1beta2.MessagingInfraPending {
+			infra.Status.Phase = v1beta2.MessagingInfraConfiguring
+		}
+		return nil
+	})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Reconcile CA
+	err = ic.ProcessSimple(func(infra *v1beta2.MessagingInfra) error {
+		return r.certController.ReconcileCa(ctx, reqLogger, infra)
+	})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Reconcile Routers
+	err = ic.ProcessSimple(func(infra *v1beta2.MessagingInfra) error {
+		return r.routerController.ReconcileRouters(ctx, reqLogger, infra)
+	})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Reconcile Brokers
+	err = ic.ProcessSimple(func(infra *v1beta2.MessagingInfra) error {
+		return r.brokerController.ReconcileBrokers(ctx, reqLogger, infra)
+	})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Update main condition
+	err = ic.ProcessSimple(func(infra *v1beta2.MessagingInfra) error {
```

Comment:
It looks like the error is not propagated.

Reply:
Thanks, will fix that.